UI/UX Improvements for the Instructor techniques setup

idegtiarov opened the pull request:
Main improvements are added to the Microscopy Technique setup while the Assignment creation.

1. Add possibility for fluent technique choosing in assignment creation flow.
2. Add Drag and Drop to the file(s) uploading process.
3. Re-design flow for microscopy analyses image choosing with and without filters.
4. Add Drag and Drop add to the image selection flow.
5. Add automation Image Set creation if more than one image is added to the certain filter at once.
6. Small fixes and improvements.

idegtiarov committed:
Implement UX dymanic technique choosing

idegtiarov committed:
Merge branch 'develop' into ux_tecnique_choice

# Conflicts:
#	instructor/common.py

idegtiarov committed:
Improve UI/UX workflow through choosing Experimental Technique

Exchange SAVE AND FINISH button from the every technique sequence with
SAVE AND CONTINUE which redirects to the common_select_technique view.

Add SAVE AND FINISH button on the common_select_technique

idegtiarov committed:
Add Drag&Drop functionality to the image loading process

Some minor changes are also added.

idegtiarov committed:
Add Drag&Drop functionality into Brightfield image selecting flow

idegtiarov committed:
Add Drag&Drop functionality into Fluorescent image selecting flow

idegtiarov committed:
Revert removed microscopy analyses and fix addSelectedImages

Adding already saved analysis images to the selected frame in the Image
Choosing window is extended with the relation to the technique name.

idegtiarov committed:
Revert: accidentally removed files

idegtiarov committed:
Small fixes and improvements

jmclaus commented:
@idegtiarov The gray highlight rectangle on the left window is overflowing:
![screenshot-pr-01](https://user-images.githubusercontent.com/1427382/32893450-cac6c124-cad9-11e7-9b7d-e1790d342294.png)

compared to:

![screenshot-prod-01](https://user-images.githubusercontent.com/1427382/32893458-d4c533d6-cad9-11e7-8076-d35103e3aa53.png)

jmclaus commented:
@idegtiarov When uploading an image using the blue `choose a file(s)` label, could the cursor turn into a pointer to signify it is clickable? Also, the `Selected Images` and `Images Bank` look like buttons but are just text column headers. They are too close to the area containing the files below them. And when hovering above them, the cursor turns into an I-beam. It should just be default.

![screenshot-pr-02](https://user-images.githubusercontent.com/1427382/32894192-5d1287b4-cadc-11e7-85f7-c3bd910a97d7.png)

jmclaus commented:
@idegtiarov The `SAVE SELECTION` and `DELETE IMAGE(S)` buttons overlap a bit the window above them.

![screenshot-pr-03](https://user-images.githubusercontent.com/1427382/32894425-0f816ab4-cadd-11e7-9096-cd85feb0017d.png)

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -408,6 +408,20 @@ $(function() {
     });
   });
+
+  /* Create new hrel URL with additionl GET prameters to reopen image chosing window */
```
Nitpick, typo, should be `choosing` not chosing.

idegtiarov replied:
fixed

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -419,35 +433,54 @@ $(function() {
         data: data
       }).then(function () {
-        window.location = '/ab/assignments/microscopy_analyze/';
+        location.href = hrefReloadUrl(); // Reload page without closing image chosing window
```
`chosing` => `choosing`

idegtiarov replied:
fixed

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -631,48 +663,107 @@ $(function() {
+   addSelectedImages(); // Add already saved images to the selected area on the page loading
   });
+
+  /* Remove remove selection from the images and revert all unsaved images to the image bank */
```
`Remove remove` => `Remove`

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -631,48 +663,107 @@ $(function() {
+      });
+  }
+
+  /* Fulfil filter area by chosen image in the Fluorescent image sececting flow*/
```
Fulfil => Fulfill and  `sececting` => `selecting`

idegtiarov replied:
fixed

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -631,48 +663,107 @@ $(function() {
+      var chosenImages = {},
+          group_id = $(".scb_ab_f_save_image").data('group_id'),
+          analysis = $(".scb_ab_f_save_image").data('analysis'),
+          filtersFrames = $(".scb_f_image_filter");
```
To follow the convention in this file of appending JQuery object vars with $, this should be named `$filtersFrames`.

idegtiarov replied:
Good catch, thanks. fixed

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -631,48 +663,107 @@ $(function() {
+          filtersFrames = $(".scb_f_image_filter");
+      if (group_id) {
+        filtersFrames.css({'visibility': 'visible', 'display': 'inline-block  '});
+        chosenImages = $(".scb_f_set[data-group_id='{}'] img".replace('{}', group_id));
```
Same here `chosenImages` => `$chosenImages`

idegtiarov replied:
fixed

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -631,48 +663,107 @@ $(function() {
+              });
+          }
+      }
+      if ($(".scb_ab_s_analyze_dialog").css('visibility') == 'visible') {
```
Please use `===` instead of `==`

idegtiarov replied:
sure thing, done

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -751,8 +842,315 @@ $(function() {
+  // Event handler for drag&drop uploading workflow
+  $filesBox.on('dragend dragover dragenter dragleave drop', function(e) {
+    e.preventDefault();
+    e.stopPropagation();
```
Could just use

`return false;`

instead of:

```
e.preventDefault();
e.stopPropagation();
```

idegtiarov replied:
improved

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -751,8 +842,315 @@ $(function() {
+
+  /* Moving selected images to the target element*/
+  function moveSelected(id, direction, target) {
+      var selectedImage = $("#" + id);
```
`selectedImage` => `$selectedImage`

idegtiarov replied:
done

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -751,8 +842,315 @@ $(function() {
+  function saveSelectedImages(group_id) {
+      var data = {};
+      data['mapping_pk'] = $(this).data('pk');
+      var filterImages = $(".scb_f_image_filter img"),
```
`filterImages` => `$filterImages`

idegtiarov replied:
done

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -751,8 +842,315 @@ $(function() {
+          url: '/ab/assignments/select_images/',
+          type: "POST",
+          data: data
+      })
```
Please add a semicolon.

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -751,8 +842,315 @@ $(function() {
+
+  $filesSelect.on('drag dragend dragover dragenter dragleave drop', function(e) {
+    e.preventDefault();
+    e.stopPropagation();
```
Please use `return false;` instead.

idegtiarov replied:
done

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -751,8 +842,315 @@ $(function() {
+
+  // Handler for the form submitting process.
+  $imageForm.on('submit', function(e) {
+      if ($imageForm.hasClass('is-uploading')) return false;
```
Please enclose `return false` in braces.

idegtiarov replied:
fixed

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -751,8 +842,315 @@ $(function() {
+
+      if ((xImage > imageX1) && (xImage < imageX2)) {
+          if ((yImage > imageY1) && (yImage < imageY2)) {
+              return true
```
Please add a semicolon.

idegtiarov replied:
done

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -751,8 +842,315 @@ $(function() {
+              return true
+          }
+      };
+      return false
```
Please add a semicolon.

idegtiarov replied:
done

jmclaus commented on instructor/ui/instructor.js:
```diff
@@ -751,8 +842,315 @@ $(function() {
+              $(this).addClass('scb_ab_s_image_selected')
+          }
+      });
+      x1 = x2 = y1 = y2 = 0
```
Please add a semicolon.

idegtiarov replied:
done

idegtiarov committed:
Fix reviews comments and requests, small additional UI fixes

jmclaus commented:
@idegtiarov The 2 symbolic links to the fonts and images directory are missing in your branch.

You can see them here:
https://github.com/starteam/starcellbio_html/tree/develop/html_app/ui

But not there:
https://github.com/starteam/starcellbio_html/tree/idegtiarov/ux_tecnique_choice/html_app/ui

Hence fonts and images are not loaded.

idegtiarov committed:
Revert symbolic links which were accidentally deleted previously

Fix small typo

idegtiarov commented:
@jmclaus thank you for paying my attention to this. I saw that accidentally deleted files, reverted them but missed that they were symbolic links.

jmclaus commented:
@idegtiarov No worries, that's why we do PR reviews! Thanks for the fix.

idegtiarov committed:
Add strain's number to the 'sample' name, fix typos

jmclaus commented on instructor/ui/assignment_setup.css:
```diff
@@ -800,7 +808,8 @@ label.custom-dropdown {
 img.scb_ab_s_library_item_wrapper{
-    width: 100px;
+    max-width: 100px;
+    max-hight: 100px;
```
@idegtiarov Typo, it should be `max-height` not `max-hight`

jmclaus commented on instructor/ui/assignment_setup.css:
```diff
@@ -904,7 +914,8 @@ input#id_objective{
     float: left;
-    width: 100px;
+    min-width: 100px;
+    min-hight: 100px;
```
Same here should be `max-height`

jmclaus replied:
@idegtiarov @laleman Checked the UI changes locally, looks better, thanks. Good to merge when above  fixes to CSS are done. :+1:

idegtiarov committed:
Hide set creation step on the fluorescent analysis image selection

jmclaus commented:
@idegtiarov Thanks for the fixes and changes. :+1:

idegtiarov committed:
Add auto scroll for the previous window position on page loading

idegtiarov committed:
Add help text and extend draggable area of the Select Image window

idegtiarov commented:
@annagav it seems PR is in a good shape to be merged?

annagav commented:
@idegtiarov @laleman When i assign an image to a filter that image disappears from the bank. Such behavior prevents the instructor from assigning the same image to different filters. Is that the desired behavior?

annagav commented:
In a case where i assigned an image to a filter and i drag another image to the same filter, that image overwrites that assignment, but the other image does not return to bank, so is not available anymore, unless i close the dialog and reopen it.

annagav commented:
The buttons are not fitting the dialog window:
![screen shot 2017-11-20 at 2 10 41 pm](https://user-images.githubusercontent.com/7574259/33036758-65b56ce6-cdfd-11e7-89a6-e72a81f30477.png)

idegtiarov commented:
> @idegtiarov @laleman When i assign an image to a filter that image disappears from the bank. Such behavior prevents the instructor from assigning the same image to different filters. Is that the desired behavior?
We have discussed this with Lourdes it is the desired behavior.

> In a case where i assigned an image to a filter and i drag another image to the same filter, that image overwrites that assignment, but the other image does not return to bank, so is not available anymore, unless i close the dialog and reopen it.
I've checked this in all browsers and cannot reproduce. After image is assinged to the filter it cannot  be overwritten

>The buttons are not fitting the dialog
I haven't check Firefox in Chrome, Safari it is ok. Fixed.

idegtiarov committed:
Fix review comment. Improve button's position.

annagav commented:
If you drag the image into the space not occupied by the image, it accepts it and overwrites the other image. 
![screen shot 2017-11-20 at 2 57 14 pm](https://user-images.githubusercontent.com/7574259/33038623-3f7b6a98-ce03-11e7-999c-7dc70dc08881.png)

idegtiarov committed:
Fix bug from with rewriting image already assign to the filter

idegtiarov commented:
@annagav  I've reproduced the bug with image rewriting you've mentioned in the comment. It was rather tricky to do )).  Oh missed your last comment with the explanation, it would have saved my time.


Fixed it. 
Other issues also fixed.

annagav commented:
looks good 👍
